common/unsafe/src/main/java/org/apache/spark/unsafe/memory/HeapMemoryAllocator.java

```diff
@@ -50,7 +50,7 @@ public MemoryBlock allocate(long size) throws OutOfMemoryError {
     long alignedSize = numWords * 8L;
     assert (alignedSize >= size);
     if (shouldPool(alignedSize)) {
-      synchronized (this) {
+      synchronized (bufferPoolsBySize) {
```

Comment:
I'm probably missing something, but how does this help? concurrent access is already excluded.

Reply:
You're right. I checked my executor log again and I find the executor got NPE first
```
java.lang.NullPointerException
	at org.apache.spark.unsafe.memory.HeapMemoryAllocator.allocate(HeapMemoryAllocator.java:58)
	at org.apache.spark.memory.TaskMemoryManager.allocatePage(TaskMemoryManager.java:302)
	at org.apache.spark.memory.MemoryConsumer.allocateArray(MemoryConsumer.java:96)
	at org.apache.spark.unsafe.map.BytesToBytesMap.allocate(BytesToBytesMap:800)
	...
```
 And later got the NoSuchElementExceptionException.
```
java.util.NoSuchElementExceptionException
        at java.util.LinkedList.removeFirst(LinkedList.java:270)
        at java.util.LinkedList.remove(LinkedList.java:685)
	at org.apache.spark.unsafe.memory.HeapMemoryAllocator.allocate(HeapMemoryAllocator.java:57)
	at org.apache.spark.memory.TaskMemoryManager.allocatePage(TaskMemoryManager.java:302)
	at org.apache.spark.memory.MemoryConsumer.allocateArray(MemoryConsumer.java:96)
	at org.apache.spark.unsafe.map.BytesToBytesMap.allocate(BytesToBytesMap:800)
	...
```
But I can't find out why NPE error here. Maybe a null WeakReference<long[]> added?
I updated the JIRA SPARK-31592.

Reply:
I don't see how it can happen. Is this on 2.4, master? the only thing added is a new WeakReference(), which can't be null.